### PR TITLE
Update documentation to cover both new and classic config formats

### PR DESCRIPTION
**Pull request by michaelklishin**
Updates the docs for http://github.com/rabbitmq/rabbitmq-server/issues/550.

### CODE REVIEW

File: site/access-control.xml [hunk 1]
```diff
@@ -297,18 +337,45 @@ limitations under the License.
...
+# rabbitmq.conf
+#
+auth_backends.1 = ldap
+# uses module name instead of a short alias, "http"
+auth_backends.2 = rabbit_auth_backend_http
+</pre>
+<pre class="example">
+%% advanced.config
+%% See HTTP backend docs for details
+{rabbitmq_auth_backend_http,
```
Comment (hairyhum) [inline site/access-control.xml]: `http.user_path`
`http.vhost_path`
`http.resource_path`
are defined in http backend plugin schema https://github.com/rabbitmq/rabbitmq-auth-backend-http/commit/571937f034fd4829ca2dfdb732de87df387c43f1

File: site/configure.xml [hunk 2]
```diff
@@ -421,10 +495,56 @@ CONFIG_FILE=/etc/rabbitmq/testdir/bunnies</pre>
            can be used to set environment variables on non-Windows
            systems.
         </p>
+        <p>
+            RabbitMQ server source repository contains
+            <a href="https://github.com/rabbitmq/rabbitmq-server/blob/stable/docs/rabbitmq.conf.example">an example rabbitmq.conf file</a>
+            named <code>rabbitmq.config.example</code>. This
```
Comment (hairyhum) [inline site/configure.xml]: Should it be `rabbitmq.conf.example`?

File: site/configure.xml [hunk 3]
```diff
@@ -421,10 +495,56 @@ CONFIG_FILE=/etc/rabbitmq/testdir/bunnies</pre>
...
+            Some configuration settings are not possible or difficult to configure
+            using the sysctl format. As such, it is possible to use an additional
+            config file in the Erlang term format (same as <code>rabbitmq.config</code>).
+            That file is commonly named <code>advanced.config</code>. It will be merged
+            with the configuration provided in <code>rabbitmq.conf</code>.
+          </p>
+          <p>
+            RabbitMQ server source repository contains
+            <a href="https://github.com/rabbitmq/rabbitmq-server/blob/stable/docs/advanced.config.example">an example advanced.config file</a>
+            named <code>rabbitmq.config.example</code>. This
```
Comment (hairyhum) [inline site/configure.xml]: Same as above comment. Should be `advanced.config.example`

File: site/configure.xml [hunk 4]
```diff
@@ -497,60 +621,62 @@ CONFIG_FILE=/etc/rabbitmq/testdir/bunnies</pre>
...
           </tr>
+          
           <tr>
             <td><code>ssl_options</code></td>
             <td>
-              SSL configuration. See the <a href="ssl.html#enabling-ssl">SSL
+              TLS configuration. See the <a href="ssl.html#enabling-ssl">TLS support
               documentation</a>.
-              <p>Default: <code>[]</code></p>
+              <p>Default: <code>none</code></p>
```
Comment (hairyhum) [inline site/configure.xml]: If `none` is defined, none of other options will work.
For example:
```
ssl_options = none
ssl_options.verify = verify_peer
```
Equivalent to 
```
ssl_options = none
```

Comment (michaelklishin) [inline site/configure.xml]: The only idea I have is to set "undefined", which can lead to exactly the same problem. Yeah, I don't think many will fall into this trap.

**Comment by hairyhum** [comment]
@michaelklishin: do we want `auth_backends`, `auth_mechanisms` and `cluster_nodes` to have `none` option?

**Comment by michaelklishin** [comment]
@hairyhum I think it would be helpful in case someone uncomments/uses the default value unintentionally.

**Comment by michaelklishin** [comment]
for `auth_backends` and `auth_mechanisms`, having no value will reject all client connections, so that probably should fail validation.

**Comment by hairyhum** [comment]
So it's better to leave it without `none`

File: site/configure.xml [hunk 5]
```diff
@@ -44,7 +44,7 @@ limitations under the License.
               set in the environment configuration file, <code>rabbitmq-env.conf</code>)
           </dd>
           <dt>
-              <a href="#configuration-file">a configuration file</a>
+              <a href="#configuration-file">One of more configuration file(s)</a>
```
Comment (camelpunch) [inline site/configure.xml]: Should this be "One or more", or maybe "One of many"?

File: site/configure.xml [hunk 6]
```diff
@@ -391,20 +393,92 @@ CONFIG_FILE=/etc/rabbitmq/testdir/bunnies</pre>
...
+          <p>
+            While the new config format is more convenient for humans to edit
+            and machines to generate, it is also relatively limited compared
+            to the classic config format used prior to RabbitMQ 3.7.0. For
+            example, when configuring <a href="/ldap.html">LDAP support</a>,
+            it may be necessary to use deeply nested data structures to
+            express desired configuration. To accommodate this need,
+            RabbitMQ still supports the classic <code>rabbitmq.config</code>
+            config files as well as ability to use both formats at the same time
+            (<code>advanced.config</code>). This is covered in more details in the following sections.
```
Comment (camelpunch) [inline site/configure.xml]: "more detail"

File: site/access-control.xml [hunk 7]
```diff
@@ -92,11 +92,17 @@ limitations under the License.
...
-          <code>[]</code>. A complete <code>rabbitmq.config</code>
+          <code>loopback_users</code> configuration to
+          <code>none</code>. A complete <a href="/configure.html">RabbitMQ config file</a>
           which does this would look like:
         </p>
         <pre class="sourcecode">
+loopback_users = none
+</pre>
+
+        Or, in the <a href="/configure.html#erlang-term-config-file">classic config file format</a> (<code>rabbitmq.config</code>):
```
Comment (dumbbell) [inline site/access-control.xml]: I would use "legacy" instead of "classic" to designate the old file formar.

File: site/access-control.xml [hunk 8]
```diff
@@ -297,18 +337,42 @@ limitations under the License.
...
+# rabbitmq.conf
+#
+auth_backends.1 = ldap
+# uses module name instead of a short alias, "http"
+auth_backends.2 = rabbit_auth_backend_http
+
+# See HTTP backend docs for details
+http.user_path = http://my-authenticator-app/auth/user
+http.vhost_path = http://my-authenticator-app/auth/vhost
+http.resource_path = http://my-authenticator-app/auth/resource
```
Comment (dumbbell) [inline site/access-control.xml]: I find the `http` prefix to be too generic in the context of the authentication backend. Maybe it should be the full module name? Or an additional prefix before the authentication mechanism name?

File: site/access-control.xml [hunk 9]
```diff
@@ -328,7 +391,17 @@ limitations under the License.
...
             IP range backend</a> for authorisation:
           </p>
-            
+
+<pre class="example">
+# rabbitmq.conf
+#
+auth_backends.1.authn = internal
+# uses module name because this backend is from a 3rd party
+auth_backends.1.authz = rabbit_auth_backend_ip_range
```
Comment (dumbbell) [inline site/access-control.xml]: I don't know enough how authentication and authorization are handled by RabbitMQ but this configuration parameters look reversed.

I would expect something like:
```
auth_backends.authn.1 = internal
auth_backends.authn.2 = http
auth_backends.authz.1 = rabbit_auth_backend_ip_range
```
In other words: RabbitMQ checks authentication first using the listed modules, then it checks authorization using another set of modules. Of course, correct me if my understanding of RabbitMQ internals is wrong.

File: site/access-control.xml [hunk 10]
```diff
@@ -358,16 +441,22 @@ limitations under the License.
       </p>
 
 <pre class="example">
+# rabbitmq.conf
+#
+auth_backends.1.authn = ldap
+auth_backends.1.authz = internal
+auth_backends.2       = internal
```
Comment (dumbbell) [inline site/access-control.xml]: Ok, so disregard my previous comment, because authentication and authorization backends are organized as pairs. Sorry for the noise!

File: site/access-control.xml (hunk 7 above)
Comment (michaelklishin) [inline site/access-control.xml]: We still support this format, and it is still used in the advanced config. It would be odd to call something legacy and then use it for both backwards compatibility and advanced cases.

File: site/access-control.xml (hunk 8 above)
Comment (michaelklishin) [inline site/access-control.xml]: I agree. Perhaps we should use module names instead (and not only in the docs).

Comment (hairyhum) [inline site/access-control.xml]: Use `auth_http` and `auth_ldap` instead `http` and `ldap` in plugin settings?

Comment (michaelklishin) [inline site/access-control.xml]: `auth_http` and `auth_ldap` sound good to me.

Comment (dumbbell) [inline site/access-control.xml]: `auth_http` is still too confusing IMHO. It could be considered a parameter of the management plugin. Using the module name makes it clear we are talking about the authentication backend.

About aliases in eg. `auth_backends.1 = http`, it's ok because it's in the context of `auth_backends`.

File: site/access-control.xml [hunk 11]
```diff
@@ -282,10 +301,31 @@ limitations under the License.
...
           <p>
+            The example above uses an alias, <code>internal</code> for <code>rabbit_auth_backend_internal</code>.
+            The following aliases are available:
+
+            <ul>
+              <li><code>internal</code> for <code>rabbit_auth_backend_internal</code></li>
+              <li><code>ldap</code> for <code>rabbit_auth_backend_ldap</code> (from the <a href="/ldap.html">LDAP plugin</a>)</li>
+              <li><code>http</code> for <code>rabbit_auth_backend_http</code> (from the <a href="https://github.com/rabbitmq/rabbitmq-auth-backend-http">HTTP auth backend plugin</a>)</li>
+              <li><code>amqp</code> for <code>rabbit_auth_backend_amqp</code> (from the <a href="https://github.com/rabbitmq/rabbitmq-auth-backend-amqp">AMQP 0-9-1 auth backend plugin</a>)</li>
+              <li><code>dummy</code> for <code>rabbit_auth_backend_dummy</code></li>
```
Comment (dumbbell) [inline site/access-control.xml]: How are aliases managed? Is it a hard-coded list? Or does RabbitMQ try to use `rabbit_auth_backend_${alias}`?

Comment (michaelklishin) [inline site/access-control.xml]: A hard-coded list.

File: site/access-control.xml (hunk 9 above)
Comment (michaelklishin) [inline site/access-control.xml]: [See this internals guide](https://github.com/rabbitmq/internals/blob/master/authorization_and_authentication.md).

File: site/access-control.xml (hunk 8 above)
Comment (michaelklishin) [inline site/access-control.xml]: Unfortunately if we go with module names in some places but not others, it will be more confusing than using just module names. Which is what we will do now.

File: site/configure.xml [hunk 12]
```diff
@@ -254,6 +257,75 @@ CONFIG_FILE=/etc/rabbitmq/testdir/bunnies</pre>
...
+                    <b>Mac OS X (Homebrew)</b> - <span class="path"><span
+                    class="envvar">${install_prefix}</span>/etc/rabbitmq/rabbitmq</span>,
+                    the Homebrew prefix is usually <code>/usr/local</code>
+                  </li>
+                  <li><b>Windows</b> - <span class="path"><span class="envvar">%APPDATA%</span>\RabbitMQ\rabbitmq</span></li>
+                </ul>
+              </td>
+              <td>
+                Main RabbitMQ config file path without the <code>.conf</code>
+                (or <code>.config</code>, for the classic format) extension.
```
Comment (dumbbell) [inline site/configure.xml]: This "without the extension" thing is so confusing. I know RabbitMQ does this for a long time, but can't we document that we require the full filename but still support the missing suffix internally?

The fact that Erlang takes a filename without its `.config` extension is a technical detail we don't need to expose to end users.

Comment (michaelklishin) [inline site/configure.xml]: @dumbbell see https://github.com/rabbitmq/rabbitmq-server/issues/691.

File: site/configure.xml [hunk 13]
```diff
@@ -484,11 +690,10 @@ CONFIG_FILE=/etc/rabbitmq/testdir/bunnies</pre>
         </doc:subsection>
 
         <doc:subsection name="config-items">
-        <doc:heading>Variables Configurable in rabbitmq.config</doc:heading>
+        <doc:heading>Commonly Used Variables Configurable in rabbitmq.conf</doc:heading>
         <p>
-          Many users of RabbitMQ never need to change any of these
-          values, and some are fairly obscure. However, for
-          completeness they are all listed here.
+          These values are configured most often. The list is not complete, as
```
Comment (dumbbell) [inline site/configure.xml]: This sentence is unclear. What do you think of:
> These variables are the most common.

File: site/configure.xml (hunk 4 above)
Comment (dumbbell) [inline site/configure.xml]: To be consistent with other examples, you should recall the parameter:
```html
<p>Default: <code>ssl_options = none</code></p>
```

File: site/configure.xml [hunk 14]
```diff
@@ -497,71 +702,87 @@ CONFIG_FILE=/etc/rabbitmq/testdir/bunnies</pre>
...
+              details and examples.
+              <p>Default: <code>listeners.ssl = none</code></p>
             </td>
           </tr>
           <tr>
-            <td><code>num_ssl_acceptors</code></td>
+            <td><code>num_acceptors.ssl</code></td>
             <td>
-              Number of Erlang processes that will accept connections for the SSL
+              Number of Erlang processes that will accept connections for the TLS
```
Comment (dumbbell) [inline site/configure.xml]: By the way, that's a good idea to replace SSL with TLS everywhere!

File: site/configure.xml [hunk 15]
```diff
@@ -883,17 +1061,19 @@ CONFIG_FILE=/etc/rabbitmq/testdir/bunnies</pre>
...
-                <li><code>pause_minority</code></li>
-                <li><code>{pause_if_all_down, [nodes], ignore | autoheal}</code>
-                  where <code>[nodes]</code> is a list of node names<br/>
-                  (ex: <code>['rabbit@node1', 'rabbit@node2']</code>)
-                </li>
                 <li><code>autoheal</code></li>
+                <li><code>pause_minority</code></li>
+                <li><code>pause_if_all_down</code></li>
+              </ul>
+              <code>pause_if_all_down</code> mode require additional parameters:
```
Comment (dumbbell) [inline site/configure.xml]: s/require/requires/

File: site/disk-alarms.xml [hunk 16]
```diff
@@ -83,12 +83,18 @@ Disabling disk free space monitoring</pre>
         required to be free on the database partition. (See the description of
         <a href="/relocate.html">file locations</a> for the default location).
         This configuration file sets the disk free space limit to 1GB:
+        <pre class="sourcecode">disk_free_limit.absolute = 1000000000</pre>
+        Or you can use memory units (KB, MB GB etc.) like this:
```
Comment (dumbbell) [inline site/disk-alarms.xml]: `kB` here? (note the lowercase `k`)

Comment (hairyhum) [inline site/disk-alarms.xml]: This format uses `cuttlefish` [bytesize](https://github.com/basho/cuttlefish/wiki/Datatypes#bytesize) datatype. So it should be KB

File: site/management.xml [hunk 17]
```diff
@@ -328,36 +334,30 @@ limitations under the License.
...
-            the <code>rabbitmq_management</code> application to the name
-            of a directory in which logs can be created and restart
-            RabbitMQ. Note that only requests to the API
-            at <code>/api</code> are logged, not requests to the static
-            files which make up the browser-based GUI.
+            Note that the definitions in the file will overwrite
+            anything already in the broker; using this option will not
+            delete anything that is already there. However, if you
+            start from a completely reset broker, use of this
+            option <b>will</b> prevent the usual default user /
```
Comment (dumbbell) [inline site/management.xml]: Please use `<strong>` instead of `<b>`.

File: site/management.xml [hunk 18]
```diff
@@ -328,36 +334,30 @@ limitations under the License.
...
+            virtual host / permissions from being created.
           </p>
-        </doc:subsection>
+        </doc:subsection>
```
Comment (dumbbell) [inline site/management.xml]: There are trailing spaces remaining here.

File: site/configure.xml [hunk 19]
```diff
@@ -497,71 +702,87 @@ CONFIG_FILE=/etc/rabbitmq/testdir/bunnies</pre>
...
-              List of ports on which to listen for AMQP connections
-              (without SSL). Can contain integers (meaning "listen on
-              all interfaces") or tuples such as <code>{"127.0.0.1",
-              5672}</code> to listen on one or more interfaces.
+              Ports or hostname/pair on which to listen for AMQP connections
+              (without TLS). See the <a href="/networking.html">Networking guide</a> for more
+              details and examples.
 
-              <p>Default: <code>[5672]</code></p>
+              <p>Default: <code>listeners.tcp.default = 5672</code></p>
```
Comment (dumbbell) [inline site/configure.xml]: The listeners configuration (for RabbitMQ/AMQP and the plugins too) is very confusing.

1. `tcp` or `ssl` below make no sense (I know it comes from the classic format). TLS is still over TCP and currently we support TCP or HTTP as the transport layer and this is orthognal to TLS. Can we use the protocol name or purpose instead (see example below)?
2. What is `default` in `listeners.tcp.default`?
3. If I compare the configuration for AMQP(S) and the management plugin:
 * `listeners` vs. `listener` (plural vs. singular)
 * `...tcp.default = 5672` vs. `...port = 15672`
 * `listeners.ssl = none` vs. `management.listener.ssl = true` (`none`/`true`)
 * `ssl_options` vs. `ssl_opts`

Below, I'm giving an example of a centralized configuration for listeners, with the following additional changes:
* I put the acceptors count and TCP options as leaves there as well.
* I use `tls` instead of `ssl` to reflect the naming used in the documentation.
* `listeners` may be too long; `listen` would be enough.

```
listen.amqp.port = 5672
listen.amqp.tls = true # Or "tls.enabled = true" if an intermediate node can't have a value
listen.amqp.tls.port = 5671
listen.amqp.tls.certfile = /path/to/cert.pem
listen.mqtt.port = 1883
listen.mqtt.tls = false
listen.mqtt.tls.port = 8883 # Ignored because TLS is disabled
listen.management.port = 15672
listen.management.tls = true
listen.management.tls.port = 15671

# For $protocol over HTTP:
listen.stomp.http.port = ...
listen.stomp.http.tls.port = ...
```

Default/common options could be set under a `default` node:
```
listen.default.nb_acceptors = 1
listen.default.handshake_timeout = 10000

listen.default.tls.enabled = true
listen.default.tls.certfile = /path/to/cert.pem
listen.default.tls.keyfile = /path/to/key.pem
listen.default.tls.nb_acceptors = 2 # Otherwise defaults to listen.default.nb_acceptors
```

This lacks an example of how to listen on specific interfaces.

What do you think?

Comment (michaelklishin) [inline site/configure.xml]: 1. TLS is over TCP but there is no word for "non-TLS" ("plain" is not widely used).
2. It's a name that can be anything.
3. Management UI probably nearly always has a single listener configured, so this is in part by design.
4. We could replace `none` with `false` but `none` means something else: it means "no options"
5. We discussed this before. If we change this to `ssl_options` it will confuse those who are used to the classic format, which is the vast majority of people. The best I can suggest is that we support both in the management plugin.

File: site/networking.xml [hunk 20]
```diff
@@ -98,6 +104,13 @@ limitations under the License.
           to listen on localhost only for both IPv4 and IPv6:
 
 <pre class="code">
+listeners.tcp.1 = 127.0.0.1:5672
+listeners.tcp.2 = ::1:5672
```
Comment (dumbbell) [inline site/networking.xml]: Is it possible to indicate an IP address without the port (so the default port is used)?

If yes, we must enclose IPv6 addresses inside square brackets (`[::1]`) like we do in URLs.

Comment (michaelklishin) [inline site/networking.xml]: Good point.

File: site/disk-alarms.xml (hunk 16 above)
Comment (michaelklishin) [inline site/disk-alarms.xml]: [We do support `kB`](https://github.com/rabbitmq/rabbitmq-server/blob/master/src/rabbit_resource_monitor_misc.erl#L43).

File: site/configure.xml (hunk 14 above)
Comment (michaelklishin) [inline site/configure.xml]: Yes but not in scope for this issue. I do it as drive-by changes whenever I work on the docs.

File: site/disk-alarms.xml (hunk 16 above)
Comment (hairyhum) [inline site/disk-alarms.xml]: We do, but cuttlefish configuration doesn't.

File: site/networking.xml [hunk 21]
```diff
@@ -105,7 +105,7 @@ Or using classic config format:
 
 <pre class="code">
 listeners.tcp.1 = 127.0.0.1:5672
-listeners.tcp.2 = ::1:5672
+listeners.tcp.2 = [::1]:5672
```
Comment (hairyhum) [inline site/networking.xml]: This should be added to example snippets in `rabbitmq_tests`

Comment (michaelklishin) [inline site/networking.xml]: I've added it and waiting for the test suite to finish.

Comment (michaelklishin) [inline site/networking.xml]: I reverted that change due to https://github.com/basho/cuttlefish/issues/216. We'll work on that later.

File: site/disk-alarms.xml [hunk 22]
```diff
@@ -83,12 +83,18 @@ Disabling disk free space monitoring</pre>
         required to be free on the database partition. (See the description of
         <a href="/relocate.html">file locations</a> for the default location).
```
Comment (bdshroyer) [inline site/disk-alarms.xml]: I'm not sure from the context here what the "default location" is referring to. Is it the free disk space limit?

Comment (michaelklishin) [inline site/disk-alarms.xml]: Default database location.
